algorithms/grandpa/prover/src/lib.rs

```diff
@@ -211,56 +212,40 @@ where
 		H::Number: finality_grandpa::BlockNumberOps,
 		T::BlockNumber: One,
 	{
-		let previous_para_hash = self
-			.para_client
-			.rpc()
-			.block_hash(Some((client_state.latest_para_height + 1).into()))
-			.await?
-			.ok_or_else(|| anyhow!("Failed to fetch previous finalized parachain + 1 hash"))?;
-		let address = parachain::api::storage().parachain_system().validation_data();
-		let validation_data = self
-			.para_client
-			.storage()
-			.at(Some(previous_para_hash))
-			.await
-			.expect("storage client should be available")
-			.fetch(&address)
-			.await
-			.unwrap()
-			.unwrap();
-		// This ensures we don't miss any parachain blocks.
-		let previous_finalized_height =
-			validation_data.relay_parent_number.min(client_state.latest_relay_height);
-
-		// we want to know if we've missed any authority set changes since we last updated the
-		// relay chain on the light client.
-		let session_end = self.session_end_for_block(client_state.latest_relay_height).await?;
-
-		if client_state.latest_relay_height != session_end && latest_finalized_height > session_end
-		{
-			latest_finalized_height = session_end
-		}
+		let mut finality_proof = if let Some(justification) = latest_justification {
+			let justification = GrandpaJustification::<H>::decode(&mut &*justification)?;
 
-		let encoded = GrandpaApiClient::<JustificationNotification, H256, u32>::prove_finality(
-			// we cast between the same type but different crate versions.
-			&*unsafe {
-				unsafe_arc_cast::<_, jsonrpsee_ws_client::WsClient>(self.relay_ws_client.clone())
-			},
-			latest_finalized_height,
-		)
-		.await?
-		.ok_or_else(|| anyhow!("No justification found for block: {:?}", latest_finalized_height))?
-		.0;
+			FinalityProof::<H> {
+				block: justification.commit.target_hash,
+				justification: justification.encode(),
+				unknown_headers: vec![],
+			}
+		} else {
+			let encoded = GrandpaApiClient::<JustificationNotification, H256, u32>::prove_finality(
+				// we cast between the same type but different crate versions.
+				&*unsafe {
+					unsafe_arc_cast::<_, jsonrpsee_ws_client::WsClient>(
+						self.relay_ws_client.clone(),
+					)
+				},
+				latest_finalized_height,
+			)
+			.await?
+			.ok_or_else(|| {
+				anyhow!("No justification found for block: {:?}", latest_finalized_height)
```

Comment:
How is it possible to have a block and no justification?

Reply:
The API returns an option so we need to handle it.